# circRNA_paperScripts
16 Nov 2020
The repository is currently under development. It contains the scripts used to generate the main figures and tables of the paper "Evolutionarily young transposable elements drive circular RNA repertoires".


## Evolutionarily young transposable elements drive circular RNA repertoires

### Franziska Gruhl, Peggy Janich, Henrik Kaessmann, David Gatfield

Circular RNAs (circRNAs) are found across eukaryotes and can function in post-transcriptional gene regulation. Their biogenesis through a circle-forming backsplicing reaction is facilitated by reverse-complementary repetitive sequences that promote pre-mRNA folding. Orthologous genes from which circRNAs arise show more conserved splice sites and exons, yet it remains unclear if this reflects genuine purifying selection acting on circRNAs or if alternative mechanisms are responsible. Here, we have determined circRNA repertoires across an evolutionary sample collection from five species representing three major mammalian lineages (marsupials, rodents, primates). Our analyses revealed that circRNA-producing loci were frequently shared across evolutionarily distant species, yet circRNA formation was associated with young, recently active and species-specific repetitive elements, rather than with shared, ancient transposon integration events. The patterns detected for circRNA parental genes indicate that circRNAs emerge as a splicing byproduct in syntenic regions serving as repeat integration hotspots. These findings argue against widespread functional circRNA conservation.
