# circRNA_paperScripts
10 Dec 2020
The repository is currently under development. It contains the scripts used to generate the main figures and tables of the paper "Evolutionarily young transposable elements drive circular RNA repertoires".

Each figure/table has its own folder with the perspective scripts and a readme-file describing how to use them. The corresponidng data files can be downloaded via [this link](https://drive.google.com/file/d/1MQt8LwJ8fpC-clU7yyt359MSMGbYE9WT/view?usp=sharing) (zipped file, same folder structure as described in the readme-files).
  

## Evolutionarily young transposable elements drive circular RNA repertoires

### Franziska Gruhl, Peggy Janich, Henrik Kaessmann, David Gatfield

Circular RNAs (circRNAs) are found across eukaryotes and can function in post-transcriptional gene regulation. Their biogenesis through a circle-forming backsplicing reaction is facilitated by reverse-complementary repetitive sequences promoting pre-mRNA folding. Orthologous genes from which circRNAs arise overall contain more strongly conserved splice sites and exons than other genes, yet it remains unclear to what extent this conservation reflects purifying selection acting on the circRNAs themselves. Here we report analyses of circRNA repertoires across five species representing three mammalian lineages (marsupials, eutherians: rodents, primates). Our analyses reveal that circRNA formation is associated with young, recently active and species-specific transposable elements, rather than with common, ancient transposon integration events, even for circRNAs formed in orthologous genes from different species. These observations suggest that circRNAs frequently emerged convergently during evolution â€“ as a byproduct of splicing in orthologs that are prone to transposable element insertion. Overall, our findings argue against widespread functional circRNA conservation.
